crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -1,402 +1,769 @@
+    }
 
-impl<T: QueryTuple> SystemParam for QuerySet<T> {
-    type Fetch = FetchQuerySet<T>;
+    /// Returns true if (and only if) this resource been mutated since the start of the frame.
+    pub fn mutated(&self) -> bool {
+        self.flags.contains(ComponentFlags::MUTATED)
+    }
+
+    /// Returns true if (and only if) this resource been either mutated or added since the start of the frame.
+    pub fn changed(&self) -> bool {
+        self.flags
+            .intersects(ComponentFlags::ADDED | ComponentFlags::MUTATED)
+    }
 }
 
-impl<'a, T: QueryTuple> FetchSystemParam<'a> for FetchQuerySet<T> {
-    type Item = QuerySet<T>;
+impl<'w, T: Component> Deref for Res<'w, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.value
+    }
+}
+
+pub struct ResState<T> {
+    component_id: ComponentId,
+    marker: PhantomData<T>,
+}
+
+impl<'a, T: Component> SystemParam for Res<'a, T> {
+    type Fetch = ResState<T>;
+}
+
+// SAFE: Res ComponentId and ArchetypeComponentId access is applied to SystemState. If this Res conflicts
+// with any prior access, a panic will occur.
+unsafe impl<T: Component> SystemParamState for ResState<T> {
+    type Config = ();
+
+    fn init(world: &mut World, system_state: &mut SystemState, _config: Self::Config) -> Self {
+        let component_id = world.initialize_resource::<T>();
+        let combined_access = system_state.component_access_set.combined_access_mut();
+        if combined_access.has_write(component_id) {
+            panic!(
+                "Res<{}> in system {} conflicts with a previous ResMut<{0}> access. Allowing this would break Rust's mutability rules. Consider removing the duplicate access.",
+                std::any::type_name::<T>(), system_state.name);
+        }
+        combined_access.add_read(component_id);
+
+        let resource_archetype = world.archetypes.resource();
+        let archetype_component_id = resource_archetype
+            .get_archetype_component_id(component_id)
+            .unwrap();
+        system_state
+            .archetype_component_access
+            .add_read(archetype_component_id);
+        Self {
+            component_id,
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Component> SystemParamFetch<'a> for ResState<T> {
+    type Item = Res<'a, T>;
 
     #[inline]
     unsafe fn get_param(
-        system_state: &'a SystemState,
+        state: &'a mut Self,
+        _system_state: &'a SystemState,
         world: &'a World,
-        _resources: &'a Resources,
-    ) -> Option<Self::Item> {
-        let query_index = *system_state.current_query_index.get();
-        *system_state.current_query_index.get() += 1;
-        Some(QuerySet::new(
-            world,
-            &system_state.query_archetype_component_accesses[query_index],
-        ))
+    ) -> Self::Item {
+        let column = world
+            .get_populated_resource_column(state.component_id)
+            .unwrap();
```

Comment:
This unwraps when asking for an uninitialized resource in a system, it should probably have a nice message (same for line 311).